### DISCUSSION TITLE
Filter hosts.

### DISCUSSION THREAD

**Discussion by opnine** [nornir-automation/nornir#750] [Q&A]
Hello, 
I have a simple script send commands, Im changing it so I choose over what devices I want to execute the script,  but I can make it work with either nr.filter or nr.inventory. 

Any help is more then welcome

thanks.

**Comment by ktbyers** [comment]
Probably need to include your code.

**Comment by opnine** [comment]
Hello sir,, 

This is my script I was trying with the nr.filter.  or nr.inventory.hosts but not sure.  

Thanks for your help.

```python
from nornir import InitNornir
from nornir_scrapli.tasks import send_command
from nornir_utils.plugins.functions import print_result
from nornir.core.filter import F  # Added(2)
import os
import json
from colorama import Fore, init

nr = InitNornir(config_file="config.yaml")

#PRINT HOST INVENTORY
print("=" * 50)
print("HOSTS IN INVENTORY")
for host in nr.inventory.hosts.keys():
    print(f"Host: {Fore.CYAN}{host}")
print(f"There are {len(nr.inventory.hosts.keys())} hosts in this inventory.")
print("=" * 50)

'''
CLEAR OS SCREEN
'''
CLEAR = "clear"
os.system(CLEAR)

'''
CHOOSE HOSTs
'''
target = input("Choose hosts:  ")
commands = txt.split(",")


'''
CHOOSE COMMANDs
'''
ctarget = input("Insert Commands:  ")
ccommands = txt.split(",")


#SEND COMMANDs

def show_commands(task):
     task.run(task=send_commands, commands=ccommands)
```

**Reply by ktbyers** [reply]
You probably need to eliminate the `show_commands` function and replace it with:

```python
nr.run(task=send_command, commands=ccommands)
```

This is at the end of your program.

Basically, your current code is never telling Nornir to execute the task (you are just defining a function which you never use).

**Reply by opnine** [reply]
I did a change , the idea is to select the hosts (target) and they apply the commands, but is not printing the outpout.
'''   
CHOOSE HOSTs
'''   
target = input("Choose hosts:  ")
target_host = nr.filter(name='target')
commands = target.split(",")
      
      
'''   
CHOOSE COMMANDs
'''   
ctarget = input("Insert Commands:  ")
ccommands = ctarget.split(",")
      
      
#SEND WITH IF
for cmd in ccommands:
      
     nr = InitNornir()
     results = target_host.run(task=send_command, commands=ccommands)
      print_result(results)
      
Thanks.

**Reply by dbarrosop** [reply]
Should be (without quotes around `target`)::

```
target_host = nr.filter(name=target)
```

**Reply by opnine** [reply]
I did the test with out the quotes and still not printing,

**Comment by dmulyalin** [comment]
Try using nornir-salt [FFun ](https://nornir-salt.readthedocs.io/en/latest/Functions/FFun.html) function, it simplifies filtering. 

Install it first - `python3 -m pip install nornir-salt`

This is sample code:
```
import pprint
from nornir import InitNornir
from nornir_salt import FFun, ResultSerializer, netmiko_send_commands

nr = InitNornir(config_file="config.yaml")

target = input("Choose hosts, comma separated list: ")
cmds = input("Commands to send, comma separated list:  ")

filtered_hosts = FFun(nr, FL=target)
output = filtered_hosts.run(task=netmiko_send_commands, commands=cmds.split(","))

results = ResultSerializer(output)

pprint.pprint(results)
```
Should print results to screen in a dictionary format keyed by show command string.

Reference and details:
FFun FL - https://nornir-salt.readthedocs.io/en/latest/Functions/FFun.html#fl-filter-list
ResultSerializer - https://nornir-salt.readthedocs.io/en/latest/Functions/ResultSerializer.html
netmiko_send_commands - https://nornir-salt.readthedocs.io/en/latest/Tasks/netmiko_send_commands.html

**Reply by opnine** [reply]
Thanks I will check it.!

**Comment by echalukyan** [comment]
Hi Team, this is just a working sample of the code: the total 2 boxes in the lab, and we filtered one by name:

```

from nornir import InitNornir
from nornir_utils.plugins.functions import print_result
from nornir_scrapli.tasks import send_command

nr = InitNornir(config_file="config.yaml")

print("=" * 50)
print("HOSTS IN INVENTORY:")
for host in nr.inventory.hosts.keys():
    print(f"Host: {host}")
print(f"There are {len(nr.inventory.hosts.keys())} hosts in this inventory.")
print("=" * 50)

# grep box with hostneme "R1"
r1 = nr.filter(name="R1")
# run task agains this box
results = r1.run(task=send_command, command="sh ip int br | ex unass")
# output in console
print_result(results)
```
